Delay subsequent requests to the same host

jNullj opened the pull request:
Closes #8350 

This adds a delay between two network downloads to avoid hitting rate limits
The delay can be set in Preferences in the RSS tab.
Added both in native GUI and webUI.
I tested this manually on Linux and i did not detect any obvious issues.

Native GUI image of the new setting:
![image](https://github.com/qbittorrent/qBittorrent/assets/15849761/b50e084b-5f8b-4194-b1db-db7593e6f413)

webUI image of the new setting:
![image](https://github.com/qbittorrent/qBittorrent/assets/15849761/aea52d72-8f39-4b1c-8cff-ebb592ad67d0)

Some other things to note which are not directly related:
1. Please note that i found the webUI to have some bugs not related to this PR, i might PR fixes later, for example changing language, and some settings not getting default value in Preferences and more...
2. Also while opening this PR i was redirected to [CODING_GUIDELINES.md](https://github.com/qbittorrent/qBittorrent/blob/35e4b35d125983051a25173d595119d99b7bda3a/CODING_GUIDELINES.md) Which encourages me to use `uncrustify`, After running it i realized many files are not formatted by it, I might get a new PR that formats everything that's missing soon

Chocobo1 commented on src/base/rss/rss_session.h:
```diff
@@ -161,6 +164,7 @@ namespace RSS
 
         CachedSettingValue<bool> m_storeProcessingEnabled;
         CachedSettingValue<int> m_storeRefreshInterval;
+        CachedSettingValue<int> m_storeFetchDelay;
```
I would suggest using a proper data type:
```suggestion
        CachedSettingValue<std::chrono::seconds> m_storeFetchDelay;
```

jNullj replied:
It looks great, more readable and safer, didn't know std had chrono.
I will add those suggestions.

Chocobo1 commented on src/base/rss/rss_session.h:
```diff
@@ -114,6 +114,9 @@ namespace RSS
         int refreshInterval() const;
         void setRefreshInterval(int refreshInterval);
 
+        int fetchDelay() const;
+        void setFetchDelay(int fetchDelay);
```
```suggestion
        std::chrono::seconds fetchDelay() const;
        void setFetchDelay(std::chrono::seconds fetchDelay);
```

Chocobo1 commented on src/gui/optionsdialog.cpp:
```diff
@@ -1190,6 +1192,7 @@ void OptionsDialog::saveRSSTabOptions() const
 
     rssSession->setProcessingEnabled(m_ui->checkRSSEnable->isChecked());
     rssSession->setRefreshInterval(m_ui->spinRSSRefreshInterval->value());
+    rssSession->setFetchDelay(m_ui->spinRSSFetchDelay->value());
```
```suggestion
    rssSession->setFetchDelay(std::chrono::seconds(m_ui->spinRSSFetchDelay->value()));
```

Chocobo1 commented on src/webui/api/appcontroller.cpp:
```diff
@@ -318,6 +318,7 @@ void AppController::preferencesAction()
 
     // RSS settings
     data[u"rss_refresh_interval"_s] = RSS::Session::instance()->refreshInterval();
+    data[u"rss_fetch_delay"_s] = RSS::Session::instance()->fetchDelay();
```
```suggestion
    data[u"rss_fetch_delay"_s] = static_cast<int>(RSS::Session::instance()->fetchDelay().count());
```

Chocobo1 commented on src/webui/api/appcontroller.cpp:
```diff
@@ -860,6 +861,8 @@ void AppController::setPreferencesAction()
     if (hasKey(u"rss_refresh_interval"_s))
         RSS::Session::instance()->setRefreshInterval(it.value().toInt());
+    if (hasKey(u"rss_fetch_delay"_s))
+        RSS::Session::instance()->setFetchDelay(it.value().toInt());
```
```suggestion
        RSS::Session::instance()->setFetchDelay(std::chrono::seconds(it.value().toInt()));
```

Chocobo1 commented on src/base/rss/rss_session.cpp:
```diff
@@ -525,6 +527,17 @@ void Session::setRefreshInterval(const int refreshInterval)
+
+void Session::setFetchDelay(const int fetchDelay)
+{
+    if (m_storeFetchDelay == fetchDelay) { return; }
```
fix coding style:
```suggestion
    if (m_storeFetchDelay == fetchDelay())  // please use getter function here
        return;
```

Chocobo1 commented on src/base/rss/rss_session.cpp:
```diff
@@ -525,6 +527,17 @@ void Session::setRefreshInterval(const int refreshInterval)
+    return m_storeFetchDelay;
+}
+
+void Session::setFetchDelay(const int fetchDelay)
```
Rename local variable to avoid name clash with `Session::fetchDelay()`:
```suggestion
void Session::setFetchDelay(const int delay)
```

Chocobo1 commented on src/webui/www/private/views/preferences.html:
```diff
@@ -731,6 +731,14 @@
+                    <label for="feed_fetch_delay">QBT_TR(Feeds fetch delay:)QBT_TR[CONTEXT=OptionsDialog]</label>
+                </td>
+                <td>
+                    <input type="text" id="feed_fetch_delay" style="width: 4em;" />&nbsp;&nbsp;QBT_TR( sec)QBT_TR[CONTEXT=OptionsDialog]
```
New code should use camel case for `id`:
```suggestion
                    <input type="text" id="feedFetchDelay" style="width: 4em;" />&nbsp;&nbsp;QBT_TR( sec)QBT_TR[CONTEXT=OptionsDialog]
```

Chocobo1 commented on src/base/rss/rss_session.cpp:
```diff
@@ -116,6 +117,7 @@ Session::Session()
     settingsStorage->removeValue(u"RSS/hosts_cookies"_s);
     settingsStorage->removeValue(u"Rss/Session/EnableProcessing"_s);
     settingsStorage->removeValue(u"Rss/Session/RefreshInterval"_s);
+    settingsStorage->removeValue(u"RSS/Session/FetchDelay"_s);
```
Remove this, you don't need it.

Chocobo1 commented on src/base/rss/rss_session.cpp:
```diff
@@ -62,6 +62,7 @@ QPointer<Session> Session::m_instance = nullptr;
 Session::Session()
     : m_storeProcessingEnabled(u"RSS/Session/EnableProcessing"_s)
     , m_storeRefreshInterval(u"RSS/Session/RefreshInterval"_s, 30)
+    , m_storeFetchDelay(u"RSS/Session/FetchDelay"_s, 2)
```
```suggestion
    , m_storeFetchDelay(u"RSS/Session/FetchDelay"_s, 2s)
```

glassez commented on src/base/net/downloadmanager.cpp:
```diff
@@ -40,11 +40,13 @@
 #include "base/global.h"
 #include "base/logger.h"
 #include "base/preferences.h"
+#include "base/rss/rss_session.h"
```
As I see it, you went the wrong way and introduced an unacceptable dependency. RSS subsystem is a client of Download Manager, so Download Manager should not depend on RSS subsystem.
If you don't have any ideas how to fix it, I can help with that later.
The first thing that comes to mind is to set a delay during sequential service registration (via `registerSequentialService()`). This will also allow you to set a separate delay for each site, if a single one for everyone is not enough (by the way, why not do it initially?).

jNullj replied:
After looking at this i think i get the issue with the dependency of Download Manager by RSS. I think using `registerSequentialService` is a great idea and can present flexibility for future growth and avoid the problematic dependency between RSS and Download Manager.

I think i could overload the function `registerSequentialService` and allow a second parameter as the delay time, then add a QHash with the delays linked to the ServiceID, a better way might be to change the ServiceID struct and add the delay to it.
Please let me know what you think so i can make another commit or a rebase to make those changes.

luzpaz commented:
ping @jNullj

jNullj commented:
> ping @jNullj

Thank you for the reminder, I may be able to fix the PR this sunday.

ASTX111 commented:
Any idea when this might be coming? It would be about the most useful update to Qbittorrent that I can think of! <3

jNullj commented:
Due to my new job, my GitHub activity has decreased. Apologies for delays. I'll aim to resume work this weekend. Feel free to co-commit or suggest changes. If maintainers feel the need to edit or create a separate PR due to my pace, I understand. Thanks for your patience.

jNullj commented:
Updated the PR:
Resolved all suggestions by @Chocobo1 at https://github.com/qbittorrent/qBittorrent/pull/19801/commits/8510d962de7a997b239bbb54488feefc5a353946 to https://github.com/qbittorrent/qBittorrent/pull/19801/commits/e1d5d28fd9f8ab2520444cf5f5ccc68029e82085
I fixed the issue spotted by @glassez 
Added code to update existing delay when the value is changed and merged latest master commits as well.

glassez commented on src/base/net/downloadmanager.cpp:
```diff
@@ -181,11 +182,17 @@ Net::DownloadHandler *Net::DownloadManager::download(const DownloadRequest &down
     m_sequentialServices.insert(serviceID);
 }
 
+void Net::DownloadManager::setSequentialServiceDelay(const ServiceID &serviceID, std::chrono::seconds delay)
```
```suggestion
void Net::DownloadManager::setSequentialServiceDelay(const ServiceID &serviceID, const std::chrono::seconds delay)
```

jNullj replied:
Resolved at a4eb01b6a8b808441f97b3e624f691945f2a4ed0

glassez commented on src/base/net/downloadmanager.cpp:
```diff
@@ -304,7 +310,10 @@ void Net::DownloadManager::processRequest(DownloadHandlerImpl *downloadHandler)
     {
-        handleDownloadFinished(downloadHandler);
+        const ServiceID id = ServiceID::fromURL(downloadHandler->url());
+        QTimer::singleShot(m_serviceDelay[id] , this, [this, id](){
+            handleDownloadFinished(id);
+        });
```
```suggestion
        QTimer::singleShot(m_serviceDelay[id] , this, [this, id]
        {
            handleDownloadFinished(id);
        });
```
or
```suggestion
        QTimer::singleShot(m_serviceDelay[id] , this, [this, id] { handleDownloadFinished(id); });
```

jNullj replied:
Resolved at 590c53dbb22e7d8f9703423259d3a167748ed56c

glassez commented on src/base/net/downloadmanager.h:
```diff
@@ -164,6 +167,7 @@ namespace Net
         QSet<ServiceID> m_sequentialServices;
         QSet<ServiceID> m_busyServices;
         QHash<ServiceID, QQueue<DownloadHandlerImpl *>> m_waitingJobs;
+        QHash<ServiceID, std::chrono::seconds> m_serviceDelay;
```
IMO, having yet another field is redundant. I would just modify existing one:
```c++
QHash<ServiceID, std::chrono::seconds> m_sequentialServices;
```

jNullj replied:
Sounds good, i made those changes at 9eb1dc23062def333d7db4e16fb21decffa878ed

glassez commented on src/base/settingsstorage.h:
```diff
@@ -84,6 +86,11 @@ class SettingsStorage final : public QObject
             const typename T::Int value = loadValue(key, static_cast<typename T::Int>(defaultValue));
             return T {value};
         }
+        else if constexpr (std::same_as<T, std::chrono::seconds>)
```
IMO, it's redundant. I would just use `CachedSettingValue<qint64>` than add all these additional types `std::chrono::seconds`, `std::chrono::milliseconds` etc. here.

jNullj replied:
I think its better to keep it that way, so it can be reused in other places that reference time because i think it might be safer to keep the types as time types and keep the conversion abstract for the developer.

I might suggest even pushing more into using chrono types, for example changing CachedSettingValue<int> for other intervals into chrono types:

example:
change m_storeRefreshInterval type to min
```cpp
        CachedSettingValue<int> m_storeRefreshInterval;
        CachedSettingValue<std::chrono::seconds> m_storeFetchDelay;
```
into
```cpp
        CachedSettingValue<std::chrono::minutes> m_storeRefreshInterval;
        CachedSettingValue<std::chrono::seconds> m_storeFetchDelay;
```

let me know if you accept or reject this so i know if i should make another update to this PR.

glassez replied:
> I might suggest even pushing more into using chrono types, for example changing CachedSettingValue for other intervals into chrono types

At least not in this PR.

jNullj replied:
I fixed as requested at https://github.com/qbittorrent/qBittorrent/pull/19801/commits/ea4ad45c85b8284efdb076e06976d05bbcda111b , I might draft a separate PR for that in the future

glassez commented on src/base/settingsstorage.h:
```diff
@@ -84,6 +86,11 @@ class SettingsStorage final : public QObject
             const typename T::Int value = loadValue(key, static_cast<typename T::Int>(defaultValue));
             return T {value};
         }
+        else if constexpr (std::same_as<T, std::chrono::seconds>)
```
IMO, it's redundant. I would just use `CachedSettingValue<qint64>` than add all these additional types `std::chrono::seconds`, `std::chrono::milliseconds` etc. here.

jNullj replied:
I think its better to keep it that way, so it can be reused in other places that reference time because i think it might be safer to keep the types as time types and keep the conversion abstract for the developer.

I might suggest even pushing more into using chrono types, for example changing CachedSettingValue<int> for other intervals into chrono types:

example:
change m_storeRefreshInterval type to min
```cpp
        CachedSettingValue<int> m_storeRefreshInterval;
        CachedSettingValue<std::chrono::seconds> m_storeFetchDelay;
```
into
```cpp
        CachedSettingValue<std::chrono::minutes> m_storeRefreshInterval;
        CachedSettingValue<std::chrono::seconds> m_storeFetchDelay;
```

let me know if you accept or reject this so i know if i should make another update to this PR.

glassez commented on src/gui/optionsdialog.ui:
```diff
@@ -3127,8 +3127,8 @@ Disable encryption: Only connect to peers without protocol encryption</string>
              <y>0</y>
-             <width>336</width>
-             <height>391</height>
+             <width>541</width>
+             <height>539</height>
```
These changes look unnecessary/redundant.

jNullj replied:
Resolved at 59c3025013775268e4e66e3509bf7b33924879b5

glassez commented on src/gui/optionsdialog.ui:
```diff
@@ -3147,14 +3147,7 @@ Disable encryption: Only connect to peers without protocol encryption</string>
                </item>
                <item>
                 <layout class="QGridLayout" name="gridLayout_5">
-                 <item row="0" column="0">
```
The changes in this file look messy. Please remake it manually to avoid unnecessary changes.

jNullj replied:
I made a new commit at 59c3025013775268e4e66e3509bf7b33924879b5 to make the diff for this PR more readable manually.
Didn't notice that Qt Designer shifted them

glassez commented on src/gui/optionsdialog.ui:
```diff
@@ -3187,18 +3186,39 @@ Disable encryption: Only connect to peers without protocol encryption</string>
+                 <item row="1" column="0">
+                  <widget class="QLabel" name="label_rss_fetch_delay">
+                   <property name="text">
+                    <string>Delay between requests:</string>
```
Wouldn't `Delay requests to same host` be more clear?

jNullj replied:
Ended up renaming it to `Same host request delay` as its a bit shorter.
at 3e91435acd2b343a8ed5c1f48c0285c509ecc1a5

jNullj commented on src/base/net/downloadmanager.cpp:
```diff
@@ -181,11 +182,17 @@ Net::DownloadHandler *Net::DownloadManager::download(const DownloadRequest &down
 }
 
-void Net::DownloadManager::registerSequentialService(const Net::ServiceID &serviceID)
+void Net::DownloadManager::registerSequentialService(const Net::ServiceID &serviceID, std::chrono::seconds delay)
```
Resolved at a4eb01b6a8b808441f97b3e624f691945f2a4ed0

jNullj commented on src/base/net/downloadmanager.cpp:
```diff
@@ -181,11 +182,17 @@ Net::DownloadHandler *Net::DownloadManager::download(const DownloadRequest &down
-void Net::DownloadManager::registerSequentialService(const Net::ServiceID &serviceID)
+void Net::DownloadManager::registerSequentialService(const Net::ServiceID &serviceID, std::chrono::seconds delay)
 {
+    setSequentialServiceDelay(serviceID ,delay);
```
Resolved at 590c53dbb22e7d8f9703423259d3a167748ed56c

glassez commented on src/base/net/downloadmanager.cpp:
```diff
@@ -182,15 +182,14 @@ Net::DownloadHandler *Net::DownloadManager::download(const DownloadRequest &down
 }
 
-void Net::DownloadManager::setSequentialServiceDelay(const ServiceID &serviceID, std::chrono::seconds delay)
+void Net::DownloadManager::setSequentialServiceDelay(const ServiceID &serviceID, const std::chrono::seconds delay)
```
It does exactly the same as `registerSequentialService()` does.

jNullj replied:
Fixed with c62f38abca02a0396b6371d2d215d0bc01c53951
I removed setSequentialServiceDelay and replace refrences with registerSequentialService.
Didn't notice QHash insert updates value if key already exists

glassez commented on src/base/net/downloadmanager.h:
```diff
@@ -164,10 +164,10 @@ namespace Net
         QNetworkProxy m_proxy;
 
-        QSet<ServiceID> m_sequentialServices;
+        // m_sequentialServices qhash value is delay for same host requests
```
```suggestion
        // m_sequentialServices value is delay for same host requests
```

glassez commented on src/gui/optionsdialog.ui:
```diff
@@ -3127,8 +3127,8 @@ Disable encryption: Only connect to peers without protocol encryption</string>
              <y>0</y>
-             <width>541</width>
-             <height>539</height>
+             <width>336</width>
```
Once again... It doesn't seem like you should to change width/height.

glassez commented on src/base/net/downloadmanager.cpp:
```diff
@@ -304,7 +309,11 @@ void Net::DownloadManager::processRequest(DownloadHandlerImpl *downloadHandler)
     connect(reply, &QNetworkReply::finished, this, [this, downloadHandler]
     {
-        handleDownloadFinished(downloadHandler);
+        const ServiceID id = ServiceID::fromURL(downloadHandler->url());
```
You need to check whether this URL is actually registered as a sequential service. At least, you shouldn't use `operator[]` below, because it will unexpectedly add this URL in sequential services registry if it wasn't there.

jNullj replied:
Good notice, Fixed with 7f6556f1703d81227435d7ef755c2bba39315d60
I added both a check for containing the key to avoid unneeded timers.
Also added default 0s value in any case.

glassez commented on src/base/rss/rss_session.cpp:
```diff
@@ -529,14 +529,14 @@ void Session::setRefreshInterval(const int refreshInterval)
 std::chrono::seconds Session::fetchDelay() const
 {
-    return m_storeFetchDelay;
+    return std::chrono::seconds(m_storeFetchDelay);
```
Wouldn't it work without repeating result type?
```suggestion
    return {m_storeFetchDelay};
```

jNullj replied:
Should work, but less explicit. I don't object as this is a short function.

glassez commented on src/base/net/downloadmanager.cpp:
```diff
@@ -304,7 +306,16 @@ void Net::DownloadManager::processRequest(DownloadHandlerImpl *downloadHandler)
+        const ServiceID id = ServiceID::fromURL(downloadHandler->url());
+        if (m_sequentialServices.contains(id))
+        {
+            QTimer::singleShot(m_sequentialServices.value(id, 0s) , this, [this, id]
```
Last comment from me.
I would put all this handling into `handleDownloadFinished` instead (since it seems to be more logically correct to delay processing the next request instead of handling previous one finishing):
```c++
void Net::DownloadManager::handleDownloadFinished(DownloadHandlerImpl *finishedHandler)
{
    const ServiceID id = ServiceID::fromURL(finishedHandler->url());
    const auto waitingJobsIter = m_waitingJobs.find(id);
    if ((waitingJobsIter == m_waitingJobs.end()) || waitingJobsIter.value().isEmpty())
    {
        // No more waiting jobs for given ServiceID
        m_busyServices.remove(id);
        return;
    }

    auto *handler = waitingJobsIter.value().dequeue();
    handler->disconnect(this);
    QTimer::singleShot(m_sequentialServices.value(id, 0s) , this, [this, handler]
    {
        qDebug("Downloading %s...", qUtf8Printable(handler->url()));
        processRequest(handler);
    }
}
```

jNullj replied:
That's a great idea, cleaner and more logical, also `handleDownloadFinished` already checks if the service is registered making it more compact.

jNullj replied:
Solved with https://github.com/qbittorrent/qBittorrent/pull/19801/commits/5722c3cd55a837292c65c3ac928fae385c7a3f81

glassez commented on src/base/rss/rss_session.cpp:
```diff
@@ -56,12 +56,14 @@ const QString DATA_FOLDER_NAME = u"rss/articles"_s;
 const QString FEEDS_FILE_NAME = u"feeds.json"_s;
 
 using namespace RSS;
+using namespace std::chrono_literals;
```
```suggestion
```

glassez commented on src/gui/optionsdialog.ui:
```diff
@@ -3155,19 +3155,22 @@ Disable encryption: Only connect to peers without protocol encryption</string>
                  </item>
                  <item row="1" column="0">
-                  <widget class="QLabel" name="label_12">
+                  <widget class="QLabel" name="label_rss_fetch_delay">
```
```suggestion
                  <widget class="QLabel" name="labelRSSFetchDelay">
```

Chocobo1 commented on src/webui/api/appcontroller.cpp:
```diff
@@ -319,6 +319,7 @@ void AppController::preferencesAction()
 
     // RSS settings
     data[u"rss_refresh_interval"_s] = RSS::Session::instance()->refreshInterval();
+    data[u"rss_fetch_delay"_s] = static_cast<int>(RSS::Session::instance()->fetchDelay().count());
```
Lets use the integer type that won't ever clip the values:
```suggestion
    data[u"rss_fetch_delay"_s] = static_cast<qlonglong>(RSS::Session::instance()->fetchDelay().count());
```

jNullj replied:
Didn't think about the edge case of users going that high with the delay.
Fixed with https://github.com/qbittorrent/qBittorrent/pull/19801/commits/aa9608600f1584fc77f3ad8d3f194b73f446ea08

Chocobo1 commented on src/webui/api/appcontroller.cpp:
```diff
@@ -865,6 +866,8 @@ void AppController::setPreferencesAction()
     if (hasKey(u"rss_refresh_interval"_s))
         RSS::Session::instance()->setRefreshInterval(it.value().toInt());
+    if (hasKey(u"rss_fetch_delay"_s))
+        RSS::Session::instance()->setFetchDelay(std::chrono::seconds(it.value().toInt()));
```
```suggestion
        RSS::Session::instance()->setFetchDelay(std::chrono::seconds(it.value().toLongLong()));
```

jNullj replied:
Fixed with https://github.com/qbittorrent/qBittorrent/commit/aa9608600f1584fc77f3ad8d3f194b73f446ea08

Chocobo1 commented on src/gui/optionsdialog.ui:
```diff
@@ -3155,19 +3155,22 @@ Disable encryption: Only connect to peers without protocol encryption</string>
+                   </property>
                    <property name="maximum">
-                    <number>2147483646</number>
+                    <number>999999</number>
```
If possible I would allow the maximum value to avoid potential user complaints about some artificial limit:
```suggestion
                    <number>2147483646</number>
```

jNullj replied:
It won't hurt anyone having the freedom to set high values. Looks like good practice to think of that case
Fixed with https://github.com/qbittorrent/qBittorrent/pull/19801/commits/6fa7c4ea81673318f80929f5fedc5bb53150de97

Chocobo1 commented on src/base/net/downloadmanager.cpp:
```diff
@@ -271,9 +274,12 @@ void Net::DownloadManager::handleDownloadFinished(DownloadHandlerImpl *finishedH
     auto *handler = waitingJobsIter.value().dequeue();
-    qDebug("Downloading %s...", qUtf8Printable(handler->url()));
-    processRequest(handler);
     handler->disconnect(this);
+    QTimer::singleShot(m_sequentialServices.value(id, 0s) , this, [this, handler]
+    {
+        qDebug("Downloading %s...", qUtf8Printable(handler->url()));
+        processRequest(handler);
+    });
```
Speculative question, should `disconnect()` happens after `processRequest()` or it doesn't matter?
```suggestion
    QTimer::singleShot(m_sequentialServices.value(id, 0s) , this, [this, handler]
    {
        qDebug("Downloading %s...", qUtf8Printable(handler->url()));
        processRequest(handler);
        handler->disconnect(this);
    });
```

glassez replied:
It is connected to be removed from waiting jobs queue in case it is destroyed before finished (e.g. when cancelled by user side). So it should be disconnected once we explicitly removed it from queue. Seems it is OK to disconnect after processRequest. The main thing is to do it before handler is deleted.

Chocobo1 reviewed (commented):
Last comments from me

glassez commented:
Unfortunately, I have found that these changes have "opened the door" to undefined behavior and/or race conditions. I want to make some preliminary changes in separate PR, and when it is merged, we will continue with this one.

glassez commented:
> I want to make some preliminary changes in separate PR

#20257

glassez commented:
@jNullj 
It looks like the place where you originally inserted the timer code was the best choice. The illogic that confused me was caused by the inappropriate naming of some things, so I corrected them in #20257. I added the timer code there in advance (with zero timeout), so all you have to do is add the code for setting the configured timeout.

glassez commented:
> should i add my (and co-authors) name/year in the files once i update them?

Your - yes.
Are there any co-authors who have made significant contributions to it? Or do you mean those who offered some tips on correcting the coding style, etc.?

jNullj commented:
> > should i add my (and co-authors) name/year in the files once i update them?
> 
> Your - yes. Are there any co-authors who have made significant contributions to it? Or do you mean those who offered some tips on correcting the coding style, etc.?

Those who offered some tips on correcting the coding style, as in those who reviewed my code and offered changes.
All changes i made which you did not offer, I created myself.

glassez commented:
> Those who offered some tips on correcting the coding style, as in those who reviewed my code and offered changes.

If the proposed changes are trivial, such as correcting the coding style, then we do not take this into account as co-authorship.

glassez commented:
@jNullj 
#20257 has just merged.

glassez commented on src/base/net/downloadmanager.cpp:
```diff
@@ -2,6 +2,7 @@
  * Bittorrent Client using Qt and libtorrent.
  * Copyright (C) 2015-2024  Vladimir Golovnev <glassez@yandex.ru>
  * Copyright (C) 2006  Christophe Dumez <chris@qbittorrent.org>
+ * Copyright (C) 2024  Jonathan Ketchker <15849761+jNullj@users.noreply.github.com>
```
1. As you might have noticed, we are adding new copyrights to the top.
2. If you don't want to provide your real email, just don't do it, you don't need to use such "noreply" address.

jNullj replied:
Fixed at 8f6218e

glassez commented:
@jNullj 
Thank you!

HanabishiRecca commented:
@glassez

HanabishiRecca reviewed (commented):
Sorry, GitHub doing some weird stuff and not published that

jNullj commented on src/webui/www/private/views/preferences.html:
```diff
@@ -2227,6 +2235,7 @@
                         // RSS Tab
                         $('enable_fetching_rss_feeds_checkbox').setProperty('checked', pref.rss_processing_enabled);
                         $('feed_refresh_interval').setProperty('value', pref.rss_refresh_interval);
+                        $('feedFetchDelay').setProperty('value', perf.rss_fetch_delay);
```
Oddly i run this branch and i just tested and able to open Advanced, but i do get other issues that don't show up as errors. Maybe that's related to running on linux?
But this should be fixed, good call.
I will push a fix.

HanabishiRecca replied:
The tab does open, but values do not load, the fields there are empty. This error was captured in Firefox, I am fairly sure JS runs the same regardless of OS.

But yeah, the typo is obviously there anyway.

jNullj replied:
I see, in js, by bad, i was looking for errors in the debug terminal, didn't realize i should look at the js as well.

jNullj replied:
Opened a PR #20326  with commit ee5a93145dfd889776fbfbb136b63056dc40d236 to solve this, as this PR is already merged
